### PR TITLE
fix: remove regex lookbehind to support more browsers

### PR DESCRIPTION
**Pull request by xushengs**
Modified regular expression to be compatible with non-V8 engine browsers, including Safari.

### CODE REVIEW

**Comment by UziTech** [comment]
That regexp is vulnerable to redos since `\\$\n` would match in `(?:\\.|[^\\\n])*?` and `(\\\$\s)+`. You can check a regexp in https://makenowjust-labs.github.io/recheck/playground/ to check if it is vulnerable to redos.

**Comment by xushengs** [comment]
Is '$ \\$ \\$ $' legal?

**Comment by UziTech** [comment]
Ya I think `$ \\$ \\$ $` should be legal

File: src/index.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import katex from 'katex';
 
-const inlineStartRule = /(?<=\s|^)\${1,2}(?!\$)/;
-const inlineRule = /^(\${1,2})(?!\$)((?:\\.|[^\\\n])+?)(?<!\$)\1(?=[\s?!\.,:]|$)/;
+const inlineStartRule = /(\s|^)\${1,2}(?!\$)/;
+const inlineRule = /^(\${1,2})(?!\$)(((?:\\.|[^\\\n])*?)((\\\$\s)|[^\$]))\1(?=[\s?!\.,:]|$)/;
```
Comment (UziTech) [inline src/index.js:4]: This would match `$\n$` which is not legal

File: src/index.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import katex from 'katex';
 
-const inlineStartRule = /(?<=\s|^)\${1,2}(?!\$)/;
-const inlineRule = /^(\${1,2})(?!\$)((?:\\.|[^\\\n])+?)(?<!\$)\1(?=[\s?!\.,:]|$)/;
+const inlineStartRule = /(\s|^)\${1,2}(?!\$)/;
+const inlineRule = /^(\${1,2})(?!\$)((?:(?:\\[^\$]|[^\\\n])*?)(?:(?:\\\$\s)+|\\n|[^\$\n]))\1(?=[\s?!\.,:]|$)/;
```
Comment (UziTech) [inline src/index.js:4]: Why is `\\n` a special case? That would also be matched by `[^\$\n]`

Comment (UziTech) [reply src/index.js:4]: I see, this actually fails if it ends with any escaped character like `$\g$`

Comment (UziTech) [reply src/index.js:4]: Also this will match a newline character after an escaped $. Newline characters should not be allowed anywhere in the inline rule

**Comment by UziTech** [comment]
I think the regexp should be `/^(\${1,2})(?!\$)((?:\\.|[^\\\n])*?(?:\\.|[^\\\n\$]))\1(?=[\s?!\.,:]|$)/`

**Comment by xushengs** [comment]
> I think the regexp should be `/^(\${1,2})(?!\$)((?:\\.|[^\\\n])*?(?:\\.|[^\\\n\$]))\1(?=[\s?!\.,:]|$)/`

This would match `$\\$$` which is not legal.

**Comment by UziTech** [comment]
Why isn't that legal? I think it should be fine

**Comment by xushengs** [comment]
Yeah, it should be fine, but there is a line in the test file: 

> 'slash $': 'must include space between katex and end delimiter: $ \\$ $'

**Comment by UziTech** [comment]
That was just a limitation to end with a `$` before not a requirement for katex

**Comment by UziTech** [comment]
This actually fixes that limitation

**Comment by xushengs** [comment]
I see!

**Review by UziTech — APPROVED** [review]
Thanks for sticking with this! 💯

**Pull request merged**